### PR TITLE
[#24789][prism] Add testdofns for later

### PR DESCRIPTION
**Pull request by lostluck**
Add the test DoFns used for bootstrap testing of prism. Unused in this PR.

These are used in the main pipeline tests for the runner, but aren't very interesting themselves.

A separate suit covers complex DoFns WRT "separation", like SDFs, process continuation and watermark estimating.

See https://github.com/apache/beam/issues/24789 for more information.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lostluck** [comment]
R: @johannaojeling 

This one is what it says on the tin. A bunch of Test DoFns that will eventually migrate into just the test file itself, instead of living in the package. 

Having them in the "production" part of the package simplifies getting coverage numbers while iterating, and ensuring the different methods on the DoFns are activated as expected. Once metrics support in the runner is sufficiently robust, that can be changed to using counters and similar.

File: sdks/go/pkg/beam/runners/prism/internal/testdofns.go [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+// CreateInitialRestriction creates the restriction being used by the SDF. In this case, the range
+// of values produced by the restriction is [Start, End).
+func (fn *selfCheckpointingDoFn) CreateInitialRestriction(_ []byte) offsetrange.Restriction {
+	return offsetrange.Restriction{
+		Start: int64(0),
+		End:   int64(10),
+	}
+}
+
+// CreateTracker wraps the fiven restriction into a LockRTracker type.
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/runners/prism/internal/testdofns.go:305]:
```suggestion
// CreateTracker wraps the given restriction into a LockRTracker type.
```

File: sdks/go/pkg/beam/runners/prism/internal/testdofns.go [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+)
+
+// The Test DoFns live outside of the test files to get coverage information on DoFn
+// Lifecycle method execution. This inflates binary size, but ensures the runner is
+// exercising the expected feature set.
+//
+// Once there's enough confidence in the runner, we can move these into a dedicated testing
+// package along with the pipelines that use them.
+
+// Registerations should happen in the test files, so the compiler can prune these
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/runners/prism/internal/testdofns.go:39]:
```suggestion
// Registrations should happen in the test files, so the compiler can prune these
```

File: sdks/go/pkg/beam/runners/prism/internal/testdofns.go [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+// This causes the restriction to be split after the claimed work and produce no primary roots.
+func (fn *selfCheckpointingDoFn) ProcessElement(rt *sdf.LockRTracker, _ []byte, emit func(int64)) sdf.ProcessContinuation {
+	position := rt.GetRestriction().(offsetrange.Restriction).Start
+
+	for {
+		if rt.TryClaim(position) {
+			// Successful claim, emit the value and move on.
+			emit(position)
+			position++
+			continue
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/runners/prism/internal/testdofns.go:338]: Is this `continue` needed?

**Review by johannaojeling — COMMENTED** [review]
LGTM!

File: sdks/go/pkg/beam/runners/prism/internal/testdofns.go (hunk 3 above)
Comment (lostluck) [reply sdks/go/pkg/beam/runners/prism/internal/testdofns.go]: Not strictly no. Removed!

**Pull request merged**